### PR TITLE
[action] Add xcodebuild_command option to scan action

### LINKED ISSUE #14861: Add sonar_build_wrapper options to run_tests action
#### Motivation Behind Feature
As an app developer, I want to have my older C/C++/Objective-C code analyzed by Sonar.  In order for the C code to be analyzed properly, SonarQube created a build wrapper executable that starts up the `xcodebuild` command.  For Swift code this build wrapper is not needed.

#### Feature Description
As a macOS/iOS app developer with older code in my apps, I would like to add a couple of optional parameters to the `run_tests` action.  The optional parameters would be

`sonar_build_wrapper` - the path to the SonarQube build-wrapper executable.  For example, "/usr/local/bin/build-wrapper-macosx-x86".

`sonar_build_wrapper_output` - the path to place the output from SonarQube's build-wrapper; required if sonar_build_wrapper is specified.

These options would take effect in scan/lib/scan/test_command_generator.rb  The updated `generate` function would look something like this:

```
parts = prefix
parts << "env NSUnbufferedIO=YES "
if config[:sonar_build_wrapper]
	parts+= "#{config[:sonar_build_wrapper]} --out-dir #{config[:sonar_build_wrapper_output]} "
end
parts += "xcodebuild"
parts += options
...
```

The new options could go into the generic build action, but it feels more natural in run_tests because Sonar can also pick up code coverage info from the same test build.

Here is an example lane using the new parameters:

 ```
 lane :test do |options|
    run_tests(
      clean: true,
      workspace: "MyApp.xcworkspace",
      scheme: "MyApp",
      devices: ["iPad (5th generation)",],
      derived_data_path: "./build/deriveddata",
      output_directory: "./build/test",
      output_types: "junit",
      code_coverage: true,
      buildlog_path: "./build/logs"
      # two new parameters here
      sonar_build_wrapper: "/usr/local/bin/build-wrapper-macosx-x86",
      sonar_build_wrapper_output: "./build/sonar_bw_output"
    )
    # Convert code coverage report to proper syntax for Sonar
    slather(
      scheme: "MyApp",
      workspace: "MyApp.xcworkspace",
      proj: "MyApp.xcodeproj",
      llvm_cov: true,
      build_directory: "./build/deriveddata",
      output_directory: "./build/test/"
    )
    # Send it to Sonar for analysis
    sonar(
      project_key: "MyAppKey",
      project_version: "1.0",
      project_name: "My App",
      sources_path: "./",
      sonar_url: "https://sonarqube.example.com/",
      sonar_login: "login_info",
      sonar_runner_args: "-Dsonar.cfamily.build-wrapper-output=./build/sonar_bw_output -Dsonar.swift.coverage.reportPath=./build/test/report.llcov -Dsonar.cfamily.llvm-cov.reportPath=./build/test/report.llcov",
      branch_name: options[:branch_name] || "local"
      )
  end
```

This will not break anything for existing users because the two new parameters to run_tests are optional.

For more info on the build wrapper see the following page:
https://docs.sonarqube.org/display/PLUG/SonarCFamily+for+Objective-C

#### Alternatives or Workarounds
The only workaround is to not use fastlane for the Sonar build.  Here is an example command:

`build-wrapper-macosx-x86 --out-dir sonar-reports xcodebuild -workspace workspace.xcworkspace -scheme scheme clean build test`

See https://community.sonarsource.com/t/objective-c-build-wrapper-with-fastlane/7837

There are no existing fastlane actions or plugins that execute the Sonar build-wrapper.

<details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                       |
| --------------------------- | ------------------------------------------- |
| OS                          | 10.14.5                                     |
| Ruby                        | 2.5.0                                       |
| Bundler?                    | false                                       |
| Git                         | git version 2.20.1 (Apple Git-117)          |
| Installation Source         | ~/.rbenv/versions/2.5.0/bin/fastlane        |
| Host                        | Mac OS X 10.14.5 (18F132)                   |
| Ruby Lib Dir                | ~/.rbenv/versions/2.5.0/lib                 |
| OpenSSL Version             | OpenSSL 1.0.2n  7 Dec 2017                  |
| Is contained                | false                                       |
| Is homebrew                 | false                                       |
| Is installed via Fabric.app | false                                       |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/ |
| Xcode Version               | 10.2.1                                      |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   |             |   |
| LANGUAGE |             |   |

### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby
before_all do |lane, options|
  ENV["FASTLANE_HIDE_GITHUB_ISSUES"]="1"
end

default_platform(:ios)

platform :ios do

  lane :test do |options|
    run_tests(
      clean: true,
      workspace: "MyApp.xcworkspace",
      scheme: "MyApp",
      devices: ["iPad (5th generation)",],
      derived_data_path: "./build/deriveddata",
      output_directory: "./build/test",
      output_types: "junit",
      code_coverage: true,
      buildlog_path: "./build/logs"
      # sonar_build_wrapper: "/usr/local/bin/build-wrapper-macosx-x86",
      # sonar_build_wrapper_output: "./build/sonar_bw_output"
    )
    # Convert code coverage report to proper syntax for Sonar
    slather(
      scheme: "MyApp",
      workspace: "MyApp.xcworkspace",
      proj: "MyApp.xcodeproj",
      llvm_cov: true,
      build_directory: "./build/deriveddata",
      output_directory: "./build/test/"
    )
    # Send it to Sonar for analysis
    sonar(
      project_key: "KEY",
      project_version: "1.0",
      project_name: "MyApp",
      sources_path: File.expand_path("../"),
      sonar_url: "https://sonarqube.example.com",
      sonar_login: "login_info",
      sonar_runner_args: "-Dsonar.cfamily.build-wrapper-output=./build/bw_output -Dsonar.swift.coverage.reportPath=./build/test/report.llcov -Dsonar.cfamily.llvm-cov.reportPath=./build/test/report.llcov",
      branch_name: options[:branch_name] || "local"
      )
  end

end

```
</details>

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.125.2 | ✅ Up-To-Date  |

### Loaded fastlane plugins:

**No plugins Loaded**

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| did_you_mean              | 1.3.0        |
| slack-notifier            | 2.3.2        |
| rouge                     | 2.0.7        |
| xcpretty                  | 0.3.0        |
| terminal-notifier         | 2.0.0        |
| terminal-table            | 1.8.0        |
| multipart-post            | 2.0.0        |
| word_wrap                 | 1.0.0        |
| public_suffix             | 2.0.5        |
| babosa                    | 1.0.2        |
| colored                   | 1.2          |
| highline                  | 1.7.10       |
| commander-fastlane        | 4.4.6        |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20180417 |
| http-cookie               | 1.0.3        |
| faraday-cookie_jar        | 0.0.6        |
| gh_inspector              | 1.1.3        |
| mini_magick               | 4.5.1        |
| multi_xml                 | 0.6.0        |
| security                  | 0.1.3        |
| xcpretty-travis-formatter | 1.0.0        |
| bundler                   | 1.17.1       |
| naturally                 | 2.2.0        |
| simctl                    | 1.6.5        |
| jwt                       | 2.1.0        |
| uber                      | 0.1.0        |
| declarative               | 0.0.10       |
| declarative-option        | 0.1.0        |
| representable             | 3.0.4        |
| retriable                 | 3.1.2        |
| mime-types                | 3.2.2        |
| multi_json                | 1.13.1       |
| signet                    | 0.11.0       |
| memoist                   | 0.16.0       |
| googleauth                | 0.6.7        |
| httpclient                | 2.8.3        |
| google-api-client         | 0.23.9       |
| nanaimo                   | 0.2.6        |
| colored2                  | 3.1.2        |
| claide                    | 1.0.2        |
| CFPropertyList            | 3.0.0        |
| atomos                    | 0.1.3        |
| xcodeproj                 | 1.9.0        |
| unicode-display_width     | 1.6.0        |
| plist                     | 3.5.0        |
| addressable               | 2.6.0        |
| tty-screen                | 0.7.0        |
| tty-cursor                | 0.7.0        |
| tty-spinner               | 0.9.1        |
| excon                     | 0.64.0       |
| faraday                   | 0.15.4       |
| unf_ext                   | 0.0.7.6      |
| fastimage                 | 2.1.5        |
| json                      | 2.2.0        |
| rubyzip                   | 1.2.3        |
| dotenv                    | 2.7.2        |
| faraday_middleware        | 0.13.1       |
| mime-types-data           | 3.2019.0331  |
| os                        | 1.0.1        |
| digest-crc                | 0.4.1        |
| google-cloud-env          | 1.0.5        |
| google-cloud-core         | 1.3.0        |
| google-cloud-storage      | 1.16.0       |
| emoji_regex               | 1.0.1        |
| openssl                   | 2.1.2        |
| date                      | 2.0.0        |
| fileutils                 | 1.2.0        |
| mini_portile2             | 2.4.0        |
| nokogiri                  | 1.10.3       |
| ipaddr                    | 1.2.2        |
| psych                     | 3.1.0        |
</details>

*generated on:* **2019-06-06**
</details>

### PR DESCRIPTION
**Pull request by mollyIV**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

> As an app developer, I want to have my older C/C++/Objective-C code analyzed by Sonar. In order for the C code to be analyzed properly, SonarQube created a build wrapper executable that starts up the xcodebuild command. For Swift code this build wrapper is not needed.

[SonarCFamily for Objective-C](https://docs.sonarqube.org/display/PLUG/SonarCFamily+for+Objective-C)

Example command line 👇 

```bash
build-wrapper-macosx-x86-64 --out-dir bw_output xcodebuild clean build
sonar-scanner
```

closes https://github.com/fastlane/fastlane/issues/14861.

### Description

Added new options to scan action: 
* ~sonar_build_wrapper~
* ~sonar_build_wrapper_output~
* xcodebuild_command

Added unit test. 

🙇 🙇

### CODE REVIEW

**Comment by mollyIV** [comment]
Unfortunately the [tests on Ubuntu](https://circleci.com/gh/fastlane/fastlane/33126?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) are failing 😬 

```
Unable to locate Xcode. Please make sure to have Xcode installed on your machine
```

@janpio may I ask you to help me on this? 🙏 The tests are not much different than the others 🤔

**Comment by janpio** [comment]
> @janpio may I ask you to help me on this? 🙏 The tests are not much different than the others 🤔

Look at the stacktrace, seems to me your addition broke some existing Xcode check somehow. Maybe this is applied onto the command after it is constructed somehow.

(Windows failure is because you have to disable all tests using Xcode - look at others how to do that, it is usually done via the test or test group)

**Comment by mollyIV** [comment]
It turned out that I needed to add a `requires_xcodebuild` param to the unit test: 

```
"uses xcodebuild_command", requires_xcodebuild: true
```

👍

File: scan/lib/scan/options.rb [hunk 1]
```diff
@@ -376,14 +376,9 @@ def self.available_options
...
-                                      env_name: "GYM_SONAR_BUILD_WRAPPER",
-                                       description: "The path to the SonarQube build-wrapper executable",
-                                       type: String,
-                                       optional: true),
-        FastlaneCore::ConfigItem.new(key: :sonar_build_wrapper_output,
-                                    env_name: "GYM_SONAR_BUILD_WRAPPER_OUTPUT",
-                                    description: "The path to place the output from SonarQube's build-wrapper, requires a sonar build wrapper",
+        FastlaneCore::ConfigItem.new(key: :xcodebuild_command,
+                                    env_name: "GYM_XCODE_BUILD_COMMAND",
+                                    description: "Allows for override of the default `xcodebuild` command",
```
Comment (janpio) [inline scan/lib/scan/options.rb]: Add the "default" here maybe so one can discover what is used by default?

File: scan/spec/test_command_generator_spec.rb [hunk 2]
```diff
@@ -106,8 +106,8 @@
 
         result = @test_command_generator.generate
         expect(result).to start_with([
-                                       "set -o pipefail && env NSUnbufferedIO=YES",
-                                       "xcodebuild",
+                                       "set -o pipefail &&",
+                                       "env NSUnbufferedIO=YES xcodebuild",
```
Comment (janpio) [inline scan/spec/test_command_generator_spec.rb]: why these test changes?

Comment (mollyIV) [reply scan/spec/test_command_generator_spec.rb]: These are outdated now I believe 🙂

File: scan/lib/scan/options.rb (hunk 1 above)
Comment (mollyIV) [reply scan/lib/scan/options.rb]: Sure, like this idea 👍 let me make the changes 👍

**Comment by mollyIV** [comment]
👋 @janpio , 

the PR is ready for another round 👀 🙏

**Review by janpio — APPROVED** [review]

**Review by joshdholtz — APPROVED** [review]
@mollyIV This change is much cleaner! Thanks for making it 😊 ❤️

**Pull request merged**